### ISSUE TITLE
32-bit data bus support

### ISSUE THREAD

**Issue by mras0** [dirkwhoffmann/Moira#8]
From https://github.com/dirkwhoffmann/vAmiga/issues/730 some thoughts:

With 020+ and MMU support coming along/planned it might be a good idea to add support for proper 32-bit bus operations sooner rather than later. If I'm reading the [source code](https://github.com/dirkwhoffmann/Moira/blob/79a292a0d4146e33bffc5caa3d3851db9081ee5b/Moira/MoiraDataflow_cpp.h#L355) correctly things are nicely prepared to support it, but all 32-bit accesses are split up into two 16-bit ones.

At the very latest, I think you'll have issues with (long)word accesses straddling protection boundaries, but I guess already with a (non-EC) 020 word access to $00ffffff you might notice a difference. Without having a complete overview, I'm guessing the "Memory" abstraction needs to be the one making the decision of whether to split up the access. Maybe it also needs to provide feedback on the access time for (more) exact emulation (at a much later point).

If/when making such changes it might be worthwhile to also consider how caches can/will be implemented. Not that I see any particular problems with the current implementation, just something to keep in mind (i.e. if 16-bit accessible chip mem is cached, you want a longword read to be read from the cache in one go, etc.)

**Comment by dirkwhoffmann** [comment]
> If I'm reading the source code correctly things are nicely prepared to support it, but all 32-bit accesses are split up into two 16-bit ones.

Yes, the current code hard-codes what the 68000 is doing. If I remember correctly, the 68020+ has two external pins for controlling the data bus width. Depending on the values on these pins, a 32-bit access is either performed as one 32-bit operation (requires a 32-bit data bus of course), two 16-bit operations, or four 8-bit operations. On my real machine, the 68030 must be forced to operate in 16-bit mode. Otherwise, it couldn't utilize the Amiga data bus.

**Comment by mras0** [comment]
> > If I'm reading the source code correctly things are nicely prepared to support it, but all 32-bit accesses are split up into two 16-bit ones.
> 
> Yes, the current code hard-codes what the 68000 is doing. If I remember correctly, the 68020+ has two external pins for controlling the data bus width. Depending on the values on these pins, a 32-bit access is either performed as one 32-bit operation (requires a 32-bit data bus of course), two 16-bit operations, or four 8-bit operations. On my real machine, the 68030 must be forced to operate in 16-bit mode. Otherwise, it couldn't utilize the Amiga data bus.

If I'm reading the manuals correctly about "Dynamic Bus Sizing" (MC68020UM §5.2.1/MC68030UM §7.2.1) you remember right. External devices can provide feedback on the buswidth (16 bits for the A500) via /DSACK{1-0}. I'm guessing that's performed by some glue logic on the board and fast RAM access is still 32-bit (or maybe that's what differentiates the TF534 from the TF536 (Z2 vs. Z3)? On [wikipedia](https://en.wikipedia.org/wiki/Amiga_2000#Variants) I can see that the "A2500" did ship with 32-bit RAM. 

Either way, this and the other "issue" were mostly meant to have a place to add comments (while they were present in my mind) for more longer term concerns that didn't fit in the vAmiga EC030 thread. As usual feel free to disregard/close at your own convenience.

**Comment by dirkwhoffmann** [comment]
Since @mithrendal has managed to implement a prototype of the 68020 instruction cache that is accurate enough to make important titles work with the AGA version of vAmiga, it's time to implement this feature properly.

Proposal: Add a function `dsack(u32 addr)` that is called before a memory access when emulating a 68020. Based on the supplied address and the emulation environment (A500, A1200, Chip RAM, Fast RAM, etc.), the client (vAmiga) returns the appropriate DSACK combination.

Moira decodes the returned DSACK signals and performs the access with the appropriate bus width (32, 16, or 8 bits). An 8-bit transfer isn't really needed for Amiga emulation, but it should be supported for completeness.

From the 68020 manual:

```
5.1.1 Dynamic Bus Sizing
The MC68020 allows operand transfers to or from 8·,16·, and 32·bit ports by dynamically
determining the port size during each bus cycle. During an operand transfer cycle, the
slave device signals its port size (byte, word, or long·word) and transfer status (complete
or not complete) to the processor through the use of the DSACKx inputs. The DSACKx in·
puts perform the same transfer acknowledge function as does the DTACK input of other
processors in the M68000 Family as well as informing the MC68020 of the current port
width. See Table 5~4 for DSACKx encodings and assertion results.
For example, if the processor is executing an instruction that requires a read of a long
word operand it will attempt to r~ad 32 bits during the first bus cycle (refer to 5.1.2 Mis·
alignment of Bus Transfers). If the port responds that it is 32 bits wide, the MC68020
latches all 32 bits of data and continues with the next operation. If the port responds that
it is 16 bits wide, the MC68020 latches the 16 bits of valid data and runs another cycle to
obtain the other 16 bits. An 8·bit port is handled similarly, but with four read cycles.
```
